File: demo/js/node.js
#!/usr/bin/node

// The node object provides support for reading messages from STDIN, writing
// them to STDOUT, keeping track of basic state, writing pluggable handlers for
// client RPC requests, and sending our own RPCs.

exports.nodeId = "";

// For console IO
var readline = require('readline');
var rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});


// STATE ///////////////////////////////////////////////////////////////////

// The local node ID
var nodeId;
// All nodes in the cluster
var nodeIds;
// Our RPC request timeout, in millis
var rpcTimeout = 1000;

// The next message ID we'll emit
var nextMsgId = 0;
// A map of RPC request message IDs to {resolve, reject} functions to handle
// their replies.
var replyHandlers = new Map();

// A map of request message types to functions which handle them.
var handlers = {};

// UTILITIES //////////////////////////////////////////////////////////////

// Our own node ID
exports.nodeId = function() {
  return nodeId;
};

// All node IDs
exports.nodeIds = function() {
  return nodeIds;
};

// Generate a new message ID
function newMsgId() {
  let id = nextMsgId;
  nextMsgId += 1;
  return id;
};
exports.newMsgId = newMsgId;

// IO /////////////////////////////////////////////////////////////////////

// Send a message body to the given node.
function send(dest, body) {
  let msg = {src: nodeId, dest: dest, body: body};
  console.warn("Sending", msg);
  console.log(JSON.stringify(msg));
};
exports.send = send;

// Reply to a request with a given response body.
function reply(req, body) {
  if (req.body.msg_id == undefined) {
    throw {code: 13,
           text: "Can't reply to request without message id: " +
                 JSON.stringify(req)};
  }
  let body2 = {... body, in_reply_to: req.body.msg_id};
  send(req.src, body2);
};
exports.reply = reply;

// Send an RPC request, returning a promise of a response which will be
// delivered a response body.
function rpc(dest, body) {
  let msgId = newMsgId();
  let body2 = {... body, msg_id: msgId};
  let promise = new Promise((resolve, reject) => {
    // Save promise delivery functions for later
    replyHandlers.set(msgId, {resolve: resolve, reject: reject});
    // Time out RPC
    setTimeout(() => {
      replyHandlers.delete(msgId);
      reject({type: 'error',
        in_reply_to: msgId,
        code: 0,
        text: 'RPC request timed out'});
    }, rpcTimeout);
  });
  // And send request
  send(dest, body2);
  return promise;
}
exports.rpc = rpc;

// Send an RPC request, and if it fails, retry it forever.
function retryRPC(dest, body) {
  return rpc(dest, body).catch((err) => {
    console.warn("Retrying RPC request to", dest, body);
    return retryRPC(dest, body);
  });
}
exports.retryRPC = retryRPC;

// REQUEST HANDLING ///////////////////////////////////////////////////////

// Register a handler for a given message type.
exports.on = function(type, handler) {
  handlers[type] = handler;
};

// Handle an init request
function handleInit(req) {
  body = req.body;
  nodeId = body.node_id;
  nodeIds = body.node_ids;
  console.warn("Node", nodeId, "initialized");
};

// Sends an error back to the client
function maybeReplyError(req, err) {
  if (body.msg_id != undefined) {
    // We can reply
    if (err.code != undefined) {
      // We have an explicitly tagged error code
      reply(req, {... err, type: 'error'});
    } else {
      // Reply with a generic error
      reply(req, {
        type: 'error',
        code: 13,
        text: String(err)
      });
    }
  }
}

// Handle a request
function handle(req) {
  try {
    let body = req.body;
    if (body.in_reply_to != undefined) {
      // This is an RPC reply; look up the promise resolve/reject functions
      let handler = replyHandlers.get(body.in_reply_to);
      // If there's no handler, this might be a duplicate message--we'll quietly
      // ignore it.
      if (handler != undefined) {
        replyHandlers.delete(body.in_reply_to);
        // If we get an error, reject the promise; otherwise deliver it.
        if (body.type === 'error') {
          handler.reject(body);
        } else {
          handler.resolve(body);
        }
      }
      return;
    }

    let type = body.type;
    if (type == "init") {
      // Special case: we call our special init handler, then a custom handler
      // if available, then ack.
      handleInit(req);
      let handler = handlers.init;
      if (handler != undefined) {
        handler(req);
      }
      reply(req, {type: 'init_ok'});
      return;
    }

    // Look up a handler for this type of message
    let handler = handlers[type];
    if (handler == undefined) {
      console.warn("Don't know how to handle msg type", type, "(", req, ")");
      reply(req, {type: 'error',
        code: 10,
        text: 'unsupported request type ' + type
      });
    } else {
      // Good, we have a handler; invoke it with our request.
      if (handler.constructor.name === 'AsyncFunction') {
        // For async handlers, attach an exception handler
        handler(req).catch((err) => {
          console.warn("Error processing async request", req, err);
          maybeReplyError(req, err);
        });
      } else {
        // Just call it directly; it'll explode.
        handler(req);
      }
    }
  } catch (err) {
    console.warn("Error processing request", err);
    maybeReplyError(req, err);
  }
};

exports.main = function() {
  rl.on('line', function(line) {
    console.warn("Got", line);
    handle(JSON.parse(line));
  });
};

File: demo/clojure/node.clj
#!/usr/bin/env bb

(deps/add-deps
  '{:deps {slingshot/slingshot {:mvn/version "0.12.2"}}})

(ns maelstrom.demo.node
  "General-purpose node operations: registering RPC handlers, sending
  messages, making our own RPC calls, and so on."
  (:require [cheshire.core :as json]
            [clojure.pprint :refer [pprint]]
            [slingshot.slingshot :refer [try+ throw+]])
  (:import (clojure.lang ExceptionInfo)
           (java.io BufferedReader)
           (java.util.concurrent CompletableFuture
                                 ExecutionException)
           (java.util.function Function)))

;; Node state

(def node-id
  "Our own node ID"
  (promise))

(def node-ids
  "All node IDs in the cluster."
  (promise))

(def next-message-id
  "What's the next message ID we'll emit?"
  (atom 0))

(def handlers
  "A map of message body types to functions which can handle those messages."
  (atom {; Ignore errors we receive when they're not in reply to anything;
         ; otherwise we'll ping-pong errors back and forth forever.
         "error" (constantly nil)}))

(def rpcs
  "A map of message IDs to Futures which should be delivered with replies."
  (atom {}))

(defn log
  "Logs a message to stderr"
  [& args]
  (locking *err*
    (binding [*out* *err*]
      (apply println args))))

(defn log-ex
  "Logs an exception (actually, any throwable), printing its stacktrace. Prints
  args first, then exception, then stacktrace."
  [^Throwable ex & args]
  (locking *err*
    (apply log args)
    (log ex)
    (.printStackTrace ex *err*)))

(defn pp-str
  "Pretty-print to a string."
  [x]
  (with-out-str (pprint x)))

;; A few utilities for futures

(defmacro then
  "Takes a CompletableFuture, a binding vector with a symbol for the value of
  that future and a body. Returns a CompletableFuture which evaluates body with
  the value bound."
  [fut [sym] & body]
  `(.thenApply ^CompletableFuture ~fut
               (reify Function
                 (apply [this# ~sym]
                   ~@body))))

(defmacro exceptionally
  "Takes a CompletableFuture, a binding vector with a symbol for an exception
  thrown by that future, and a body. Returns a CompletableFuture which
  evaluates body with that exception bound, if one is thrown."
  [fut [sym] & body]
  `(.exceptionally ^CompletableFuture ~fut
                   (reify Function
                     (apply [this# ~sym]
                       ~@body))))

;; Working with node IDs

(defn other-node-ids
  "A set of other node IDs."
  []
  (disj (set @node-ids) @node-id))

;; Core node functions

(defn add-handler!
  "Registers a handler. Takes a message body type (e.g. :echo) and a
  function of a request, which should be invoked whenever an RPC request of
  that type arrives."
  [type handler]
  (swap! handlers assoc (name type) handler))

(defmacro defhandler
  "Sugar for add-handler!. Takes an RPC type, a docstring, a binding
  form of one argument (the request), and a body to evaluate. Defines a
  function to handle that RPC and calls it whenever a request arrives."
  [type docstring bindings & body]
  (let [fn-name (gensym (str "handle-" type))]
    `(do (defn ~fn-name ~docstring ~bindings ~@body)
         (add-handler! '~type ~fn-name))))

(defn send!
  "Sends a message on stdout."
  [dest body]
  (locking *out*
    (println (json/generate-string {:src @node-id, :dest dest, :body body}))))

(defn reply!
  "Replies to a request message with the given body."
  [req body]
  (send! (:src req) (assoc body :in_reply_to (:msg_id (:body req)))))

(defn rpc!
  "Sends an RPC request body to the given node, and returns a CompletableFuture
  of a response body."
  [dest body]
  (let [fut (CompletableFuture.)
        id  (swap! next-message-id inc)]
    (swap! rpcs assoc id fut)
    (send! dest (assoc body :msg_id id))
    fut))

(defn handle-reply!
  "Handles a reply to an RPC we issued."
  [{:keys [body] :as reply}]
  (when-let [fut (get @rpcs (:in_reply_to body))]
    (if (= "error" (:type body))
      (.completeExceptionally fut (ex-info (:text body)
                                           (dissoc body :type :text)))
      (.complete fut body)))
  (swap! rpcs dissoc (:in_reply_to body)))

(defn handle-init!
  "Handles an init message by saving the node ID and node IDS to our local
  state."
  [{:keys [body] :as req}]
  (deliver node-id (:node_id body))
  (deliver node-ids (:node_ids body))
  (log "I am node" @node-id)
  ; Also try a custom handler if one's been registered.
  (when-let [handler (get @handlers "init")]
    (handler req))
  (reply! req {:type :init_ok}))

(declare handle-txn!)

(defn handle-req!
  "Handles incoming request messages"
  [{:keys [body] :as req}]
  (let [type (:type body)]
    (if (= type "init")
      (handle-init! req)
      (if-let [handler (get @handlers type)]
        (handler req)
        (throw (ex-info (str "Unknown request type " type) {:code 10}))))))

(defn process-stdin!
  "Mainloop which handles messages from stdin"
  []
  (doseq [line (line-seq (BufferedReader. *in*))]
    (future
      (let [req (json/parse-string line true)]
        (try
          (if (:in_reply_to (:body req))
            (handle-reply! req)
            (handle-req! req))
          (catch ExceptionInfo e
            ; Send these back to the client as error messages
            (reply! req (assoc (ex-data e)
                               :type :error
                               :text (ex-message e))))
          (catch Exception e
            (locking *err*
              (log "Error processing request" req)
              (log e))
            ; And send a general-purpose error message back to the client
            (reply! req {:type :error
                         :code 13
                         :text (ex-message e)})))))))

(defn start!
  "Starts running the node."
  []
  (try
    (process-stdin!)
    (catch Throwable t
      (log-ex t "Fatal error in mainloop!"))))
